### PR TITLE
Add http basic auth as an option (with tests)

### PR DESCRIPTION
**Pull request by alext**
## What

This supersedes #16, to take that implementation and add tests around it.

I've also taken the opportunity to refactor it a little and to add some tests around the construction of the http server in the main package.

## How to review

* Code review
* Deploy the exporter and verify it works.
* Enable basic auth, and verify that works.

## Who can review

Not me.

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			req := httptest.NewRequest("GET", "http://www.example.com/metrics", nil)
+			req.SetBasicAuth(username, "not-the-password")
+			resp := httptest.NewRecorder()
+			server.Handler.ServeHTTP(resp, req)
+
+			Expect(resp.Code).To(Equal(http.StatusUnauthorized))
+			Expect(promHandler.ReceivedRequests()).To(HaveLen(0))
+		})
+
+		It("allows vlaid requests through", func() {
```
Comment (tlwr) [inline main_test.go]: typo

**Comment by tlwr** [comment]
Spotted a typo, directly commented

I agree with all your changes

I would `s/auth_username/authUsername/` and likewise for `auth_password`

I am assuming that your intention is to be able to run with only `AUTH_USERNAME` set and with `AUTH_PASSWORD` being optional (i.e. `""`)

File: main_test.go (hunk 1 above)
Comment (alext) [reply main_test.go]: Fixed...

**Comment by alext** [comment]
> I would s/auth_username/authUsername/ and likewise for auth_password

Doh, yes indeed. Fixed...

**Comment by tlwr** [comment]
Request a review from Phil because he has _extremely_ more context

**Comment by philandstuff** [comment]
I don't have anything to add other than what I said on the other PR. I'm no go expert so I can't judge the quality of the added tests

**Pull request merged**